File: go/cmd/aoc2024/day04.go
package main

import (
	"aoc/aoc"
	"fmt"
	"strconv"
)

type day04 struct {
	// horribly assuming that I do not have to deal with UTF8 in the input
	puzzle []string
	width  int
	height int
	part1  int
	part2  int
}

func Day04() aoc.Solver {
	return &day04{
		puzzle: make([]string, 0),
	}
}

func (solver *day04) Parse(line string) {
	// straight line or matrix?
	solver.puzzle = append(solver.puzzle, line)
	solver.height += 1
	if solver.width == 0 {
		solver.width = len(line)
	} else if solver.width != len(line) {
		panic(fmt.Sprintf("Unexpected line lenght: got %d, expected %d", len(line), solver.width))
	}
}

func (solver *day04) Solve() (*string, *string) {
	for row := 0; row < len(solver.puzzle); row++ {
		for col := 0; col < len(solver.puzzle[row]); col++ {
			if solver.puzzle[row][col] == 'X' {
				checkPart1(solver, row, col)
			} else if solver.puzzle[row][col] == 'A' {
				checkPart2(solver, row, col)
			}
		}
	}
	part1 := strconv.Itoa(solver.part1)
	part2 := strconv.Itoa(solver.part2)
	return &part1, &part2
}

func checkPart1(solver *day04, row int, col int) {
	// navigate all directions
	for _, dir := range aoc.ALL_SURROUNDING {
		start := aoc.RowColToGridPos(col, row)
		if checkDir(solver, start, dir) {
			solver.part1 += 1
			aoc.Info("[%d] Found word at %v -> %v", solver.part1, start, dir)
		}
	}
}

var directions = []aoc.GridPos{aoc.MOVE_DL, aoc.MOVE_DR, aoc.MOVE_UR, aoc.MOVE_UL}

var checks = []string{
	// M.M
	// .A.
	// S.S
	"MMSS",
	// M.S
	// .A.
	// M.S
	"MSSM",
	// S.S
	// .A.
	// M.M
	"SSMM",
	// S.M
	// .A.
	// S.M
	"SMMS",
}

func checkPart2(solver *day04, row int, col int) {
	center := aoc.RowColToGridPos(col, row)
	for _, letters := range checks {
		matches := 0
		for i, dir := range directions {
			next := center.Add(dir)
			if !next.InBounds(solver.width, solver.height) {
				break
			}
			if solver.puzzle[next.Row][next.Col] != letters[i] {
				break
			}
			matches++
		}
		if matches < 4 {
			continue
		}
		// if all match we have a winner
		solver.part2 += 1
		aoc.Info("[%d] Found X-MAS at %v", solver.part2, center)
		return
	}
}

const SEARCH = "MAS"

func checkDir(solver *day04, start aoc.GridPos, dir aoc.GridPos) bool {
	// star is already correct
	next := start
	for i := range SEARCH {
		next = next.Add(dir)
		if !next.InBounds(solver.width, solver.height) {
			return false
		}
		if solver.puzzle[next.Row][next.Col] != SEARCH[i] {
			return false
		}
	}
	return true
}

File: csharp/year2025/Day04.cs
using adventofcode.utils;

namespace adventofcode.year2025;

public class Day04 : ISolver
{
    private long _part1;
    private long _part2;
    private int _width;
    private readonly List<char[]> _map = new();
        
    public void Parse(string input)
    {
        if (_width != 0 && _width != input.Length)
        {
            throw new Exception($"Invalid input unmatched width: {_width}/{input.Length}: {input}");
        }
        _width = input.Length;
        _map.Add(input.ToCharArray());
    }

    public (string? part1, string? part2) Solve()
    {
        while (true)
        {
            var removable = Enumerable
                .Range(0, _map.Count)
                .SelectMany(row => Enumerable.Range(0, _width)
                    .Where(col => _map[row][col] == '@')
                    .Select(col => (row, col, GridPos.AllSurrounding
                        .Select(dir => new GridPos(col, row).Add(dir))
                        .Count(next =>
                            next.InBounds(_width, _map.Count) &&
                            _map[next.Row][next.Col] == '@'))))
                .Where(tup => tup.Item3 < 4)
                .ToList();
            if (removable.Count == 0)
            {
                break;
            }
            if (_part1 == 0)
            {
                _part1 = removable.Count;
            }
            _part2 += removable.Count;
            // remove all removable
            foreach (var (row, col, _) in removable)
            {
                _map[row][col] = '.';
            }
        }
            
        return (_part1.ToString(), _part2.ToString());
    }
}